transportable-udfs-trino-plugin/src/main/java/com/linkedin/transport/trino/TransportConnectorFactory.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright 2019 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.transport.trino;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Module;
+import io.airlift.log.Logger;
+import io.trino.spi.connector.Connector;
+import io.trino.spi.connector.ConnectorContext;
+import io.trino.spi.connector.ConnectorFactory;
+import io.trino.spi.function.FunctionId;
+import java.io.File;
+import java.io.FileFilter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.*;
+
+
+public class TransportConnectorFactory implements ConnectorFactory {
+
+  private static final String DEFAULT_TRANSPORT_UDF_REPO = "/transport-udf-repo";
+  private static final String TRANSPORT_UDF_REPO_CONFIG_NAME = "transport.udf.repo";
+  private static final Logger log = Logger.get(TransportConnectorFactory.class);
+
+  private static final FileFilter TRANSPORT_UDF_JAR_FILTER = (file) -> {
+    return file.isFile() && file.getName().endsWith(".jar") && !file.getName().startsWith("transportable-udfs");
+  };
+
+  private Connector connector;
+  private final Class<? extends Module> module;
+
+
+  // test only
+  public TransportConnectorFactory(Connector connector) {
+    this.connector = connector;
+    this.module = TransportModule.class;
+  }
+
+  public TransportConnectorFactory(Class<? extends Module> module) {
+    this.connector = null;
+    this.module = module;
+  }
+
+  @Override
+  public String getName() {
+    return "TRANSPORT";
+  }
+
+  @Override
+  public Connector create(String catalogName, Map<String, String> config, ConnectorContext context) {
+    requireNonNull(config, "config  is null");
+    if (this.connector == null) {
+      ClassLoader classLoaderForFactory = TransportConnectorFactory.class.getClassLoader();
+      List<URL> jarUrlList = getUDFJarUrls(config);
+      TransportUDFClassLoader classLoaderForUdf = new TransportUDFClassLoader(classLoaderForFactory, jarUrlList);
+      ServiceLoader<StdUdfWrapper> serviceLoader = ServiceLoader.load(StdUdfWrapper.class, classLoaderForUdf);
+      List<StdUdfWrapper> stdUdfWrappers = ImmutableList.copyOf(serviceLoader);
+      Map<FunctionId, StdUdfWrapper> functions = new HashMap<>();
+      for (StdUdfWrapper wrapper : stdUdfWrappers) {
+        log.info("Loading Transport UDF class: " + wrapper.getFunctionMetadata().getFunctionId().toString());
+        functions.put(wrapper.getFunctionMetadata().getFunctionId(), wrapper);
+      }
+      this.connector = new TransportConnector(functions);
```

Comment:
this will cache the connector, returning the same instance if `create()` is called multiple times. is this the expected behavior?

Reply:
`create()` should only be called once at server startup when connector is loaded so this is the expected behavior. In other plugins it is usually handled by injection.

Reply:
if `create()` should only be called once, then why would we cache the value?

> In other plugins it is usually handled by injection.

So why don't we use injection?

Reply:
@yiqiangin Let's re-model this with injection or remove the cache. You can refer to [example-http-plugin](https://github.com/trinodb/trino/tree/414/plugin/trino-example-http/src/main/java/io/trino/plugin/example) for reference

In `TransportModule` you can have
```
binder.bind(TransportConnector.class).in(Scopes.SINGLETON);
```
then create a `TransportConfig` class to load the user-provided `transport-udf-repo`
```
public class TransportConfig
{
    private String transportUdfRepo;

    @Config("transport-udf-repo")
    public TransportConfig setTransportUdfRepo(String path)
    {
        this.transportUdfRepo = path;
        return this;
    }
}
```

Reply:
Done with dependency injection